Implement reinforcement heuristic

AntonOellerer opened the pull request:
For selecting good territories for placing units, we implement a heuristic which first searches for a continent, where the relation of total troops to our troops is not too high, but still favourable, and then we scan our territories for a territory having a favourable relation of troops in it to enemy troops around it.

Commit:
Implement placement heuristic

For selecting good territories for placing units, we
implement a heuristic which first searches for a continent,
where the relation of total troops to our troops is not too
high, but still achievable, and then we scan our territories
for a territory having a similar relation of troops in it to
enemy troops around it.

Commit:
Implement placement heuristic

For selecting good territories for placing units, we
implement a heuristic which first searches for a continent,
where the relation of total troops to our troops is not too
high, but still achievable, and then we scan our territories
for a territory having a similar relation of troops in it to
enemy troops around it.

Commit:
Merge branch 'feature/reinforcement' of github.com:AntonOellerer/risk_crusher into feature/reinforcement

Commit:
Improve & fix algorithm

* Fix calculation of neighbouring enemy troops reducing
    territory id
* Use difference between enemy & own troop relation to find
    best territory
* Use `.orElseGet` to make code more functional for cases
    where we don't find a fitting continent
* Filter out continents where we don't have troops at the start
    instead of after the relation calculation
* Reinforce strongest instead of weakest territory in case of
    not finding something good
* Correct territory relation calculation to prevent division
    by 0

Commit:
Calculate and place the maximum amount of troops

Commit:
Add fix to play card when needed

freecraver commented on src/main/java/agents/leeroy/Leeroy.java:
```diff
@@ -66,11 +71,109 @@ private void setPhase(Risk game) {
+        return RiskAction.select(initialPlacementRoot.getId());
+    }
+
+    private RiskAction reinforce(Risk game) {
```
would probably make sense to extract this into a dedicated class. I'd like to transitition to a plug-in approach (e.g. with behavior-patterns or smth like that) which would allow us more easily to compare different combinations against each other. Let's discuss tomorrow :)

freecraver commented on src/main/java/agents/leeroy/Leeroy.java:
```diff
@@ -66,11 +71,109 @@ private void setPhase(Risk game) {
+                .collect(Collectors.groupingBy(RiskTerritory::getContinentId,
+                        Collectors.summingDouble(RiskTerritory::getTroops)));
+
+        var continentEstimatedBest = Stream
```
quite hard to read 🧐

Commit:
Fix reinforcing

There were some additional errors in the reinforcement logic.

Commit:
Merge branch 'main' into feature/reinforcement

# Conflicts:
#	src/main/java/agents/leeroy/Leeroy.java

Commit:
Further work on Reinforce

* Move the reinforcement logic to a specific class
* Use Pair instead of Abstract Map Entry Set for pair passing
* Deal with case where there is a clean cut between continent
    ownerships
* Introduce fallback for when we have to place troops in
    certain territories due to card trade in

Commit:
Modify thresholds to favor almost conquered continents

AntonOellerer commented:
@freecraver can I merge this?

AntonOellerer commented:
I think I noticed a weakness in the current algorithm:
If we have to trade in cards in the phase, we do all of the reinforcement before we trade the cards in, so the additional reinforcements go to less important territories. 
I would merge the branch for now and deal with this later though tbh.

Commit:
Trade in cards as soon as necessary

Until now, we only traded cards in when the engine forced us
to. This meant that we did all of the placing before
receiving the additional troops, and needing to place them
on suboptimal countries.
This commit fixes this by detecting if we need to trade in
cards, and doing so as soon as possible.

Commit:
Merge branch 'main' into feature/reinforcement

# Conflicts:
#	src/main/java/agents/leeroy/Leeroy.java

Commit:
Merge main into feature/reinforcement

AntonOellerer commented:
I actually managed to quickly implement a heuristic to trade in cards as soon as necessary
